### PR TITLE
[gazebo] new port

### LINKED ISSUE #8014: [New Port Request] Gazebo
Library name: Gazebo

Library description: A physics and robot simulation framework. It is near ubiquitous in the world of robotics research.

Direct from their website:
"Robot simulation is an essential tool in every roboticist's toolbox. A well-designed simulator makes it possible to rapidly test algorithms, design robots, perform regression testing, and train AI system using realistic scenarios. Gazebo offers the ability to accurately and efficiently simulate populations of robots in complex indoor and outdoor environments. At your fingertips is a robust physics engine, high-quality graphics, and convenient programmatic and graphical interfaces. Best of all, Gazebo is free with a vibrant community."

Source repository URL: https://bitbucket.org/osrf/gazebo/src/default/

Project homepage (if different from the source repository): gazebosim.org

Anything else that is useful to know when adding (such as optional features the library may have that should be included):  Gazeo depends on some of the "ignition" libraries: https://ignitionrobotics.org/home. Ignition Math, Ignition Msgs, and Ignition Transport are required as dependencies (possibly others).

### PR DESCRIPTION
**Pull request by Ace314159**
**Describe the pull request**

- #### What does your PR fix?  
  Fixes #8014

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  all, Yes

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  `Yes`

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

**Comment by JonLiu1993** [comment]
@Ace314159 ,Thanks for your pr, if this pr ready for review, please let me know

**Comment by Ace314159** [comment]
Sorry for the delay. Although the build passes, there are still some unexpected runtime issues making the tools unusable, which I am currently trying to fix. I will mark the PR as ready for review once I fix them.

**Comment by Ace314159** [comment]
Just to provide an update, I'll need to wait for this upstream PR (https://github.com/osrf/gazebo/pull/3161) to be merged before this can be ready.

**Review by JonLiu1993 — APPROVED** [review]

**Comment by JonLiu1993** [comment]
@Ace314159 ,Have you tested the features locally?

**Comment by Ace314159** [comment]
@JonLiu1993 Yeah, I have tested building and using all the features and no features. However, I haven't tested all the combinations since there are a lot of combinations.

**Review by JonLiu1993 — APPROVED** [review]

**Comment by mathisloge** [comment]
I'm wondering whether the port name should be versioned. Since gazebo itself is just called gazebo it might be a good thing to call the port gazebo and use the normal vcpkg versioning.

**Comment by Ace314159** [comment]
I've just been following the naming scheme of the other OSRF ports, such as [sdformat9](https://github.com/microsoft/vcpkg/blob/master/ports/sdformat9), [sdformat10](https://github.com/microsoft/vcpkg/tree/master/ports/sdformat10), [ignition-msgs5](https://github.com/microsoft/vcpkg/tree/master/ports/ignition-msgs5), [ignition-msgs6](https://github.com/microsoft/vcpkg/tree/master/ports/ignition-msgs6), etc.

I think @traversaro would be a better person to ask regarding why that was done and if gazebo should follow that convention as well.

**Comment by ras0219-msft** [comment]
Including the version number in the package name is the right approach if you can expect to use the different packages at the same time. For example, if they install differently named libraries and different headers.

Based on my reading of the source[1], this library just produces a `libgazebo` which would conflict with some future "gazebo12". Therefore, I think this library should be just `gazebo`. This matches what Debian does, for example[2] (our packages match their 'source packages').

[1] https://github.com/osrf/gazebo/blob/gazebo11/gazebo/CMakeLists.txt#L107-L112
[2] https://packages.debian.org/source/sid/gazebo

**Comment by traversaro** [comment]
@ras0219-msft is right. While ignition robotics libraries (including the new "Ignition Gazebo") can be installed side-by-side (hence the reason why the major number is included in the package name in vcpkg, debian, conda-forge), for "Classic Gazebo" this is not the case, so it probably make sense to just call it `gazebo`.

**Comment by Ace314159** [comment]
I see that makes sense! I've renamed it

File: ports/gazebo/vcpkg.json [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        "cpp"
+      ]
+    },
+    "ignition-common3",
+    "ignition-fuel-tools4",
+    "ignition-math6",
+    "ignition-msgs5",
+    "ignition-transport8",
+    "ogre",
+    "pkgconf",
```
Comment (ras0219-msft) [inline ports/gazebo/vcpkg.json:33]: This should be a host dependency, otherwise it may not be executable on the current machine (imagine cross-compiling to android).

File: ports/gazebo/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        simbody   CMAKE_DISABLE_FIND_PACKAGE_Simbody
+        dart      CMAKE_DISABLE_FIND_PACKAGE_DART
+        bullet    CMAKE_DISABLE_FIND_PACKAGE_BULLET
+        libusb    NO_LIBUSB_FEATURE
+        gdal      NO_GDAL_FEATURE
+        graphviz  NO_GRAPHVIZ_FEATURE
+)
+
+vcpkg_add_to_path("${CURRENT_HOST_INSTALLED_DIR}/debug/bin")
+vcpkg_add_to_path("${CURRENT_HOST_INSTALLED_DIR}/bin")
```
Comment (ras0219-msft) [inline ports/gazebo/portfile.cmake:26]: Why are these needed? I don't see any host dependencies that would provide any `/bin` contents, so these should do nothing.

```suggestion
```

File: ports/gazebo/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+    OPTIONS
+        -DUSE_EXTERNAL_TINY_PROCESS_LIBRARY=ON
+        -DPKG_CONFIG_EXECUTABLE=${CURRENT_INSTALLED_DIR}/tools/pkgconf/pkgconf.exe
```
Comment (ras0219-msft) [inline ports/gazebo/portfile.cmake:31]:
```suggestion
        -DPKG_CONFIG_EXECUTABLE=${CURRENT_HOST_INSTALLED_DIR}/tools/pkgconf/pkgconf.exe
```

File: ports/gazebo/0001-Fix-deps.patch [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
++      libswscale_FOUND AND FFMPEG_FEATURE)
+     set (HAVE_FFMPEG TRUE)
+   else ()
+     set (HAVE_FFMPEG FALSE)
+@@ -538,7 +538,7 @@ if (PKG_CONFIG_FOUND)
+   ########################################
+   # Find GNU Triangulation Surface Library
+   pkg_check_modules(gts gts)
+-  if (gts_FOUND)
++  if (gts_FOUND AND GTS_FEATURE)
```
Comment (ras0219-msft) [inline ports/gazebo/0001-Fix-deps.patch:37]:
```suggestion
+  if (GTS_FEATURE)
```

File: ports/gazebo/0001-Fix-deps.patch [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
++if (NOT GDAL_FOUND OR NO_GDAL_FEATURE)
+   message (STATUS "Looking for libgdal - not found")
+   BUILD_WARNING ("GDAL not found, Digital elevation terrains support will be disabled.")
+   set (HAVE_GDAL OFF CACHE BOOL "HAVE GDAL" FORCE)
+@@ -718,7 +719,7 @@ endif()
+ ########################################
+ # Find graphviz
+ include (${gazebo_cmake_dir}/FindGraphviz.cmake)
+-if (NOT GRAPHVIZ_FOUND)
++if (NOT GRAPHVIZ_FOUND OR NO_GRAPHVIZ_FEATURE)
```
Comment (ras0219-msft) [inline ports/gazebo/0001-Fix-deps.patch:80]:
```suggestion
+if (NO_GRAPHVIZ_FEATURE)
```

File: ports/gazebo/0001-Fix-deps.patch [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
++  if (NOT libusb-1.0_FOUND OR NO_LIBUSB_FEATURE)
+     BUILD_WARNING ("libusb-1.0 not found. USB peripherals support will be disabled.")
+     set (HAVE_USB OFF CACHE BOOL "HAVE USB" FORCE)
+   else()
+@@ -648,7 +649,7 @@ endif ()
+ ########################################
+ # Find gdal
+ include (FindGDAL)
+-if (NOT GDAL_FOUND)
++if (NOT GDAL_FOUND OR NO_GDAL_FEATURE)
```
Comment (ras0219-msft) [inline ports/gazebo/0001-Fix-deps.patch:71]:
```suggestion
+if (NO_GDAL_FEATURE)
```

File: ports/gazebo/0001-Fix-deps.patch [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
++    set(BULLET_VERSION ${BULLET_VERSION_STRING})
+     add_definitions( -DLIBBULLET_VERSION=${BULLET_VERSION} )
+   else()
+     set (HAVE_BULLET FALSE)
+@@ -572,7 +573,7 @@ if (PKG_CONFIG_FOUND)
+   ########################################
+   # Find libusb
+   pkg_check_modules(libusb-1.0 libusb-1.0)
+-  if (NOT libusb-1.0_FOUND)
++  if (NOT libusb-1.0_FOUND OR NO_LIBUSB_FEATURE)
```
Comment (ras0219-msft) [inline ports/gazebo/0001-Fix-deps.patch:62]:
```suggestion
+  if (NO_LIBUSB_FEATURE)
```

**Review by ras0219-msft — CHANGES_REQUESTED** [review]
The vcpkg understanding of the dependencies must be correct -- so if we're adding options like `NO_GRAPHVIZ_FEATURE`, they should replace any existing logic.

File: ports/gazebo/portfile.cmake (hunk 2 above)
Comment (Ace314159) [reply ports/gazebo/portfile.cmake:26]: This is required because the build process uses the `protoc` tool from protobuf, which needs the protobuf DLLs to be in the path. I tried just adding the protobuf package bin to the path, but it didn't work in the CI.

Here's an example of error it gets:
```
cmd.exe /C "cd /D D:\vcpkg\buildtrees\gazebo\x64-windows-dbg\gazebo\msgs && D:\vcpkg\installed\x64-windows\tools\protobuf\protoc.exe --plugin=protoc-gen-gazebomsgs=D:/vcpkg2/buildtrees/gazebo/x64-windows-dbg/gazebo/msgs/gazebomsgs_out.exe --cpp_out=dllexport_decl=GZ_MSGS_VISIBLE:D:/vcpkg2/buildtrees/gazebo/x64-windows-dbg/gazebo/msgs --gazebomsgs_out=D:/vcpkg2/buildtrees/gazebo/x64-windows-dbg/gazebo/msgs --proto_path=D:/vcpkg2/buildtrees/gazebo/src/46e867c51d-e170e5c60c.clean/gazebo/msgs D:/vcpkg2/buildtrees/gazebo/src/46e867c51d-e170e5c60c.clean/gazebo/msgs/collision.proto"
--gazebomsgs_out: protoc-gen-gazebomsgs: Plugin failed with status code 3221225781.
```

I was able to find the solution using https://github.com/osrf/gazebo/issues/2932

**Comment by Ace314159** [comment]
Tensorflow seems to be failing due to a dead link

**Review by JonLiu1993 — APPROVED** [review]

**Comment by JonLiu1993** [comment]
@Ace314159 ,Have you tested the features locally?

**Comment by Ace314159** [comment]
Yeah, like before, I've tested with no features and all features.

**Review by JonLiu1993 — APPROVED** [review]

**Comment by JonLiu1993** [comment]
@ras0219-msft , Could you please take a look?

File: ports/gazebo/portfile.cmake (hunk 2 above)
Comment (Ace314159) [reply ports/gazebo/portfile.cmake:26]: Perhaps they shouldn't be in the host installed directory but just normal installed directory?

**Comment by Ace314159** [comment]
@ras0219-msft I'd love if you could review my PR once again. Thank you!

File: ports/gazebo/portfile.cmake (hunk 2 above)
Comment (vicroms) [reply ports/gazebo/portfile.cmake:26]: What about cross-compilation? I assume that `gazebomsgs_out.exe` is created as part of the port build, it would need appropriate DLLs when building for x86 for example.

Comment (ras0219-msft) [reply ports/gazebo/portfile.cmake:26]: This will need to be the normal installed directory assuming that the plugin is being built as part of this build. That design prevents cross-compilation, which is highly unfortunate.

Normally, our suggestion to enable cross-compilation here would be to add a way to export the plugin executable from the host build and consume it from target builds, then add a dependency from the target build onto the host build. However, gazebo's _enormous_ dependency list (including all of Qt & ffmpeg) makes the prospect of building once for the host _and again_ for the target quite daunting.

The ideal solution would be to split off the gazebomsgs plugin into a separate port which could be built for the host and then consumed by the main build. I assume that plugin has few-to-no dependencies other than protobuf. However that's a very high bar of build system surgery to do as part of the vcpkg packaging and I wouldn't block the PR over that.

Comment (Ace314159) [reply ports/gazebo/portfile.cmake:26]: @ras0219-msft Thanks for the detailed explanation regarding cross compilation!

Since it looks like `ignition-msgs5` currently doesn't support cross compiling in vcpkg, what do you propose I should do?

**Review by vicroms — APPROVED** [review]
